### ISSUE TITLE
blacklist for servers a player does not wish to see again

### ISSUE THREAD

**Issue by Sokomine** [minetest/minetest#3068]
Feature request: There are sometimes servers that are very disappointing and where the rules are incompatible with the ones the player likes. It ought to be possible to blacklist them so that they don't show up in the list anymore.

**Comment by RHRhino** [comment]
Good idea :+1:

**Comment by C1ffisme** [comment]
Yes. Especially if you don't want your kids going on the really bad ones, like the ones where there are no rules.

**Comment by 4aiman** [comment]
As for the the feature itself - it's rather trivial to implement. 
The thing is, how "secure" this should be?
Even the best cryptography is powerless against the ```Del``` button...


I'm not against, but even 7 y.o. child will be able to read in the minetest.config.example the name of a file with banned servers and delete that.

I'd say it's parents responsibility to look after their children and not let them be 1 on 1 with a PC/notebook/whatever device. 
Or else...
http://www.youtube.com/watch?v=YDTC2_TKyvw
https://www.youtube.com/watch?v=1SkfcofStKs
(Search for dubbed source on your own, please...)

**Comment by C1ffisme** [comment]
You could put it in a hidden folder with a name like ```guilists.lua``` or something so that it's not suspicious.

**Comment by 4aiman** [comment]
@C1ffisme 
Can't tell when this one have started to involve myself...

Anyway,
One *can't* do that.
*Somewhere* a path to that file *should* be stored.
I doubt it should be hardcoded.
There's an option to *not* mention that file in config example... 
But then again, how's anyone going to become aware of that setting? ;)


**My concern is rather simple:**
As a feature added to let users to remove the servers from the public list for whatever reason *is* a bright idea.


But no blacklist should be treated like a way to stop someone using some/any servers.
If someone wishes to ultimately ban some server it would be better to block it *using the features provided by the installed OS*  or any 3rd-party SW, but not Minetest.

**Comment by Sokomine** [comment]
Just to clarify: I want a strictly local blacklist so that servers I personally don't like don't show up for *me* anymore. Think of it like deleting a bookmark, or of removing a tv station from the sender list if you're not intrested in the program. The short names of the servers and the random sorting of a large list make it difficult to recognize servers. This is a problem for servers that showed up positive as well. Many inexperienced players will not find them again if they're not on the top of the list. The favorites are of limited help here.

**Comment by 4aiman** [comment]
@Sokomine :+1: 
I think likewise.

However, to blacklist a server one would have to login and find out the one is bad...


**Related (TMM):**

>  Many inexperienced players will not find them again if they're not on the top of the list.

That's why I made one public list of a two in Magichet.
Favourites first, then the full list of servers:
![screenshot - 20 08 2015 - 11 06 22](https://cloud.githubusercontent.com/assets/2913210/9378625/89cfe346-472b-11e5-8acd-2fdadd789a5a.png)
^ This helps me and many others to have their favourites at the top of a list.

There's the :star: button to add selected public server to the favourites list.
To remove some public server from favourites one will need to click that :star: button again.

One can notice that there are 2 "Xanadu" servers - one is marked with a :star: and the other is not.
The one on the top is from the favourites list.


There's also the ```        Add         ``` button which helps to add custom servers which are not in the public list. 
Added servers go straight to the favourites list.
I'd dare to say that's a neat feature to have if one doesn't have any "Internets" but would like to host a LAN game :)

And if there's no Internet connection then only the "favourites" part of a list will appear.
There won't be much use of the servers that has been added by adding a public server to the favourites list, but having the list of manually added servers is smth that just should be there.


Moreover, I've expanded a bit the amount of info about servers that is to be saved.
That enabled me to show server flags and the slots count in the favourites list.
Those can change, true, but it's really nice to let the game show you what type of server it (at least) was. 

When I want to go all creative I don't have to recollect which server from my favourites *is* creative - the game shows me that much and saves my traffic and time.

**Comment by Sokomine** [comment]
Marking favored servers with a star and combining both lists is also a good idea. Perhaps with an additional sort option? Blacklisted servers would go to the very end of the list and get a - as a mark in the local list. The amount of times a server has been visited is also something that might help a lot in the list.

**Comment by NathanSalapat** [comment]
I agree, a blacklist would be helpful. There are many servers that I'm not interested in playing on, due to language differences, French, German, etc, or servers that I just dislike because of the mods and rules, or lack thereof.

**Comment by hellomickey22** [comment]
@4aiman I WANT TO MAKE A REVIEW ABOUT A BEAUTIFUL SERVER, BUT WITH HORRIBLE AND EXCESSIVE MODERATORS: XANADU. YOU CANT FIGHT, YOU CANT BUILD YOUR NICE BUILDING. YOU CANT DO ANYTHING. JUST BE VICTIMS OF MODERATORS (ISKA, ICEAGECOMING, SKROTE, BONNY, NOOB101 AND GLORFINDEL). if you build near at a road, besides of a desert, in area that wanst claimed by them, they will remove your buildings very soon. They'll use any arguments for kick the user. https://forum.minetest.net/viewtopic.php?f=10&t=8494&p=214321#p214321

**Comment by DonBatman** [comment]
@hellomickey22 this is a discussion about minetest. This is not a place to talk trash about servers. There are many servers that you can play on. Minetest dev team does not control the servers.
Please use the forum for your grievances.

**Comment by hellomickey22** [comment]
@DonBatman Where's a forum that can I talk about my experience on that server?

**Comment by rubenwardy** [comment]
https://forum.minetest.net/viewtopic.php?f=10&t=8494

**Comment by C1ffisme** [comment]
Well... Seeing as this issue was awakened, I suppose we could bring it up for discussion again.

**Comment by NathanSalapat** [comment]
I would still like a way to "favorite" servers, and to make servers not show, or show at the bottom. I know I can turn off the public server list to only see servers that I have already connected to, but I think a simple favorite button, would be a nice addition.

**Comment by C1ffisme** [comment]
What about a whilelist of servers? We could make Minetest turn into 'kid mode' and the only servers that would show up would be whitelisted ones. A smarter kid might find a way to bypass kid mode, but this would keep the younger ones, maybe 7 or 8 year olds from getting on servers that one would not want their kids on.

The setting to show only whitelisted servers would be in `minetest.conf`, it would be named something like `show_only_whitelisted_servers`.

As another security measure, there should be a `singleplayer_only` option, in case you do not want kids to get on any server.

--------------------

Or, if it needs to be a blacklist, we could have the file require root privileges to edit, although this would be a drastic measure.

Another idea is to hash the blacklist like passwords, so if you log into a server, it would hash it. If the hashed server address matches one of the blacklisted ones, then you cannot log on.

Of course, a blacklist isn't a very good idea, since there will always be new servers with the same bad content. I think a white list is the way to go.

**Comment by NathanSalapat** [comment]
Oh, I like the idea of a singleplayer_only option.

**Comment by DonBatman** [comment]
Having a single player only setting would be good. Parents of younger children would love it.

I like the blacklist idea but it should be button in the server menu to remove servers. I don't think the dev team should be the ones who decide which servers are family friendly and which are not. Plus this way people could remove servers for other reasons. They may not like creative servers so they could remove those, for example.

A simple "remove from list" button would be the way to go imho.

**Comment by ExeterDad** [comment]
I like that idea very much. My kids aren't allowed to enable the server lists. I would rather be less restrictive and open their doors a bit after I've culled out servers I don't want them stumbling upon.

**Comment by C1ffisme** [comment]
> I like the blacklist idea but it should be button in the server menu to remove servers

@DonBatman Oh yeah. It would be. But the problem is that some kids are real smart, and know how to delete the blacklist file, or something like that.

> They may not like creative servers so they could remove those, for example.

I think that the ability to filter servers would be good.

Another idea is to have a rating for each server, such as G, PG, PG-13, etc. (Hopefully not R. Unless that server gets... really out of hand. I'm thinking anarchy servers.) The only problem with that, as I've seen on the Minecraft Forum Server Lists, is that people often just leave it at 'G' (I guess that is the default option). We also might have problems with people deciding what is PG and what is PG-13. (Example : One person might think that it's okay to say 'crap' on a PG server, while another might think that PG means that it mustn't have someone saying 'dang it' on it. Or someone might think that PG means PG from the 1980's, which a lot of stuff from then that was PG would probably be PG-13 nowadays.)

**Comment by rubenwardy** [comment]
It's already technically possible to make a currated server list using https://github.com/rubenwardy/mt_curation

Just change this setting:

    serverlist_url = minetest.rubenwardy.com/curated

There's only one server there at the moment.

**Comment by C1ffisme** [comment]
@rubenwardy That's great, but it still doesn't fulfill everybody's needs, and as you said, only has one server at the moment.

(P.S. This isn't relevant to blacklists, but the mods in your README should probably be links to the mods' forum posts, since some people don't know what those mods are, and thus what mods would be similar.)

**Comment by rubenwardy** [comment]
I'm talking about the software, not the list instance as such. That's why I said technically possible.

**Comment by C1ffisme** [comment]
@rubenwardy Ah, I see.

**Comment by 4aiman** [comment]
Now everyone can change just one setting and apply for a free censorship.
Nice one, rubenwardy! :-1:

**Comment by C1ffisme** [comment]
@4aiman Forgive me for not seeing how bad of an idea this is, because I can't see any catches right now.

**Comment by Calinou** [comment]
It makes sense to have a personal blacklist just for yourself, to hide low-quality servers from your own server list view. Most Valve games like Team Fortress 2 already have such an option.

**Comment by rubenwardy** [comment]
> Now everyone can change just one setting and apply for a free censorship.

Saying parents should not be allowed to show a subset of the serverlist?

**Comment by rubenwardy** [comment]
> It makes sense to have a personal blacklist just for yourself, to hide low-quality servers from your own server list view.

I agree that this is a better way of doing it. I was just saying that it is already possible to make a whitelist.

**Comment by kilbith** [comment]
Worked on a new favorite system recently...

![capture d ecran de 2016-04-18 14-07-03](https://cloud.githubusercontent.com/assets/7883281/14603626/244b2654-056f-11e6-95c6-3ee9dd5743ae.png)

However the mainmenu code is so crappy I'm not sure I'll finish it...

**Comment by 0-afflatus** [comment]
Can we make parental controls a separate issue?
My understanding of Sokomine's OP is about players being able to filter their own choices.
Discussing censorship here is irrelevant and confusing.

**Comment by paramat** [comment]
hellomickey22 deleted your abusive and offtopic comment.

**Comment by paramat** [comment]
Any comments? Anyone interested in coding this sometime? If not closing in 1 month.

**Comment by C1ffisme** [comment]
@paramat Wait... What did hellomickey... Nevermind. I won't ask.

Anyways, I feel like the earlier point about a 7 year-old deleting a blacklist was brought up earlier. I feel like there are three possible answers to this:

1- Do what rubenwardy did, and provide a list of bad servers to blacklist (or good servers to whitelist) online.

2- Don't show a list of detected servers at all, rather, the IPs of safe servers can be found on the Minetest Forums or another trustworthy list and entered into Minetest.

3- Just leave it up to the parent? I feel like teaching your kids not to click on unknown servers in the serverlist is a life lesson on the same level as teaching them not to watch random videos on Youtube.

**Comment by paramat** [comment]
:-1: 

No core dev interest and disapproved. So 'PRs may be accepted' is unsuitable.
